pkg/utils/utils.go

```diff
@@ -0,0 +1,173 @@
+	// update request URL path
+	req.URL.Path = t.Path
+
+	// populate proto, namespace, service, and port to request headers
+	req.Header.Set("Cluster-Proxy-Proto", t.Proto)
+	req.Header.Set("Cluster-Proxy-Namespace", t.Namespace)
+	req.Header.Set("Cluster-Proxy-Service", t.Service)
+	req.Header.Set("Cluster-Proxy-Port", t.Port)
+
+	return req
+}
+
+// GetTargetServiceConfig extrict the target service config from requestURL
+// input: https://<route location cluster-proxy>/cluster1/api/v1/namespaces/default/services/<https:helloworld:8080>/proxy-service/ping?time-out=32s
+// output: TargetServiceConfig{Cluster: cluster1, Proto: https, Service: helloworld, Namespace: default, Port: 8080, Path: /ping}
+func GetTargetServiceConfig(requestURL string) (ts TargetServiceConfig, err error) {
+	urlparams := strings.Split(requestURL, "/")
+	if len(urlparams) < 9 {
+		err = fmt.Errorf("requestURL format not correct, path less than 9: %s", requestURL)
+		return
+	}
+
+	namespace := urlparams[5]
+
+	proto, service, port, valid := utilnet.SplitSchemeNamePort(urlparams[7])
+	if !valid {
+		return TargetServiceConfig{}, fmt.Errorf("invalid service name %q", urlparams[7])
+	}
+	if proto == "" {
+		proto = "https" // set a default to https
+	}
+
+	servicePath := strings.Join(urlparams[9:], "/")
+	servicePath = strings.Split(servicePath, "?")[0] //we only need path here, the proxy pkg would add params back
+
+	return TargetServiceConfig{
+		Cluster:   urlparams[1],
+		Proto:     proto,
+		Service:   service,
+		Namespace: namespace,
+		Port:      port,
+		Path:      servicePath,
+	}, nil
+}
+
+// GetTargetServiceConfigForKubeAPIServer extrict the kube apiserver config from requestURL
+// input: https://<route location cluster-proxy>/cluster1/api/pods?timeout=32s
+// output: TargetServiceConfig{Cluster: cluster1, Proto: https, Service: kubernetes, Namespace: default, Port: 443, Path: api/pods}
+func GetTargetServiceConfigForKubeAPIServer(requestURL string) (ts TargetServiceConfig, err error) {
+	ts = TargetServiceConfig{
+		Proto:     "https",
+		Service:   "kubernetes",
+		Namespace: "default",
+		Port:      "443",
+	}
+
+	paths := strings.Split(requestURL, "/")
+	if len(paths) <= 2 {
+		err = fmt.Errorf("requestURL format not correct, path more than 2: %s", requestURL)
+		return
+	}
+	kubeAPIPath := strings.Join(paths[2:], "/")      // api/pods?timeout=32s
+	kubeAPIPath = strings.Split(kubeAPIPath, "?")[0] // api/pods note: we only need path here, the proxy pkg would add params back
+
+	ts.Cluster = paths[1]
+	ts.Path = kubeAPIPath
+	return ts, nil
+}
+
+// GetTargetServiceURLFromRequest is used on the agent side, the service-proxy agent recived a request from the proxy-agent, and need to know the target service URL to do further proxy.
+func GetTargetServiceURLFromRequest(req *http.Request) (*url.URL, error) {
+	// get proto, namespace, service, and port from request headers
+	proto := req.Header.Get("Cluster-Proxy-Proto")
```

Comment:
These headers need to be declared as constant with a comment for each one. It is not clear what is the purpose.